provd/internal/testutils/gdm_mock.go

```diff
@@ -0,0 +1,222 @@
+package testutils
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"log/slog"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/canonical/ubuntu-desktop-provision/provd/internal/consts"
+	"github.com/godbus/dbus/v5"
+)
+
+var (
+	privateBusMU sync.Mutex
+	privateBuses []*privateBus
+)
+
+type privateBus struct {
+	conn    *dbus.Conn
+	tempDir string
+	address string
+	cleanup func()
+}
+
+type gdmManagerBus struct{}
+
+func (g gdmManagerBus) OpenSession() (interface{}, *dbus.Error) {
+	bus, err := setupPrivateBus()
+	if err != nil {
+		return "", dbus.NewError(consts.DbusGdmPrefix+".Manager.OpenSession", []interface{}{"failed to open session"})
+	}
+
+	// Add the bus to the list of private buses so it can be cleaned up later
+	privateBusMU.Lock()
+	defer privateBusMU.Unlock()
+	privateBuses = append(privateBuses, bus)
+	return bus.address, nil
+}
+
+// ExportGdmMock exports a mock GDM Manager on the system bus.
+func ExportGdmMock(conn *dbus.Conn) error {
+	g := gdmManagerBus{}
+	if err := conn.Export(g, dbus.ObjectPath("/org/gnome/DisplayManager/Manager"), consts.DbusGdmPrefix+".Manager"); err != nil {
+		return fmt.Errorf("could not export GDM Manager mock: %w", err)
+	}
+
+	reply, err := conn.RequestName(consts.DbusGdmPrefix, dbus.NameFlagDoNotQueue)
+	if err != nil {
+		return fmt.Errorf("failed to acquire account name on local system bus: %w", err)
+	}
+	if reply != dbus.RequestNameReplyPrimaryOwner {
+		return fmt.Errorf("failed to acquire account name on local system bus: name already taken")
+	}
+
+	return nil
+}
+
+type gdmSessionBus struct {
+	conn     *dbus.Conn
+	username string
+}
+
+func (g *gdmSessionBus) Ping() *dbus.Error {
+	return nil
+}
+
+func (g *gdmSessionBus) BeginVerificationForUser(serviceName, username string, s dbus.Sender) (interface{}, *dbus.Error) {
+	// Save the username for later verification
+	g.username = username
+	_ = emitSignal(g.conn, dbus.ObjectPath("/org/gnome/DisplayManager/Session"), string(s), consts.DbusGdmPrefix+".UserVerifier.SecretInfoQuery", "gdm-password", "foobar")
+	return "", nil
+}
+
+func (g *gdmSessionBus) AnswerQuery(serviceName, answer string, s dbus.Sender) *dbus.Error {
+	if g.username != "myuser" || answer != "mypassword" {
+		_ = emitSignal(g.conn, dbus.ObjectPath("/org/gnome/DisplayManager/Session"), string(s), consts.DbusGdmPrefix+".UserVerifier.Problem", "gdm-password", "Invalid username or password")
+	}
+	_ = emitSignal(g.conn, dbus.ObjectPath("/org/gnome/DisplayManager/Session"), string(s), consts.DbusGdmPrefix+".Greeter.SessionOpened", "gdm-password")
+	return nil
+}
+
+func (g *gdmSessionBus) StartSessionWhenReady(serviceName string, shouldStartSession bool) *dbus.Error {
+	return nil
+}
```

Comment:
3: ...and in your mocks return errors when those flags are set to simulate those cases